### ISSUE TITLE
DOMWidget messages sent from Python to Javascript are lost before widget is displayed.

### ISSUE THREAD

**Issue by jordoncm** [ipython/ipywidgets#300]
Any messages sent using `self.send` are lost if the widget has not yet been displayed in the notebook.

``` javascript
%%javascript

require(['widgets/js/widget', 'widgets/js/manager', 'jquery', 'underscore'], function(widget, manager, $, _) {
    var MyCustomDOMWidget = widget.DOMWidgetView.extend({
        initialize: function() {
            this.listenTo(
                this.model,
                'msg:custom',
                function() { console.log(arguments); });
        }
}
```

``` python
from ipywidgets import widgets
import traitlets

class MyCustomDOMWidget(widgets.DOMWidget):
    _view_name = traitlets.Unicode('MyCustomDOMWidget', sync=True)
```

``` python
from IPython.display import display

foo = MyCustomDOMWidget()
foo.send('bar')  # This message is lost to the ages.
display(foo)
foo.send('baz')  # This message is received, though this may have to be in a second cell.
                 # May also get lost due to DOM render delay.
```

This is really problematic for me as I am mapping a few methods across from Python to Javascript and they only work if they are called in right order in following cells.

**Comment by SylvainCorlay** [comment]
@jordoncm at the point where you send the first message, your widget does not have a view yet.
I think that you might actually want to create a custom widget *model* that is instantiated as soon as the python-side widget is instantiated and handles this type of message.

**Comment by jordoncm** [comment]
@SylvainCorlay You are saying I could extend https://github.com/ipython/ipywidgets/blob/master/ipywidgets/static/widgets/js/widget.js#L41 to queue messages or something?

I am not sure what you mean.

**Comment by SylvainCorlay** [comment]
That is exactly right. Then on the python side, you specify the `_model_name` attribute in the same way as you did for the `_view_name`.

Within ipywidgets, there are some examples sur as the `Box` model, and the JavaScript Link model. In custom widgets libraries like bqplot, there are other examples of custom models.

**Comment by jordoncm** [comment]
I took a bit of a different approach; rather than queuing messages in the model for the view, I decided to just have the model `listenTo` `msg:custom` in `initialize` and then add the payloads of the message to a Baclbone.Collection as a property on the model. Then the view when it is created just stays in sync with changes to the collection property on the model. This seemed a little more proper MVC to me.

I was able to get it working but I noticed two other issues:

I am not able to use `defaults` to create Javascript side properties for a model and I cannot use `this.set` inside of `initialize` to set the initial value.

``` javascript
%%javascript

require(['widgets/js/widget', 'widgets/js/manager', 'jquery', 'underscore'], function(widget, manager, $, _) {
    var Layer = Backbone.Model.extend({
        defaults: function() {
            return {
                config: {},
                type: undefined,
                visible: true
            };
        }
    });

    var LayerCollection = Backbone.Collection.extend({
        model: Layer
    });

    var MyCustomModel = widget.WidgetModel.extend({
        // This will cause an error.
        defaults: function() {
            return {
                layers: new LayerCollection()
            };
        },

        initialize: function() {
            this.listenTo(
                this, 'msg:custom', _.bind(this.handleMessage, this));
            // This will also cause an error.
            this.set('layers', new LayerCollection());
        },

        handleMessage: function(payload) {
            // This is how I was able to work around this issue, by waiting 
            // until I needed it to initialize.
            /*
            if (!this.get('layers')) {
                this.set('layers', new LayerCollection());
            }
            */
            switch(payload.action) {
                case 'addLayer':
                    this.get('layers').add({
                        config: payload.config,
                        type: payload.type
                    });
                    break;
            }
        }
    });

    var MyCustomDOMWidget = widget.DOMWidgetView.extend({
        ...
    }
});
```

I don't have the error handy but it seemed like attributes was undefined in each case (very similar errors) so it couldn't add properties.

**Comment by SylvainCorlay** [comment]
Well, you should not need to "queue" the messages for the view in the model. The model in ipywidgets is a MVC, where the role of the view would be rather minimal (only rendering of the model state). You want to be able to display another view of that same model at any point in time.

**Comment by SylvainCorlay** [comment]
Closing the issue for now since it is not a bug.

**Comment by jordoncm** [comment]
I am not sure there is not a bug here. You are right I do not need to queue messages if I just maintain model state instead. However I seem pretty unable to create a simple Backbone.Collection property on the widget model. In my opinion having to wait to set it a first use is bad; since `defaults` and `this.set` in `initialize` don't work (see my second example).

**Comment by SylvainCorlay** [comment]
@jordoncm are you running ipywidgets from source or the last release?

**Comment by SylvainCorlay** [comment]
Ah, yes, it is a legitimate bug but it is fixed in master. cf #266.

**Comment by jordoncm** [comment]
Good to hear. I am pretty sure I am on the latest release. I think Jupyter notebook just updated a day or so ago. I am running the version before that.

**Comment by SylvainCorlay** [comment]
If you want to run ipywidgets master from source, please use the latest release of notebook (the one from today), which is required.

ipywidgets 5.0 should be released in a couple of weeks.

**Comment by evert2410** [comment]
Your example has caused me a lot of headaches :-), and raised a question

_initialize: function() {
            this.listenTo(
                this, 'msg:custom', _.bind(this.handleMessage, this));
            // This will also cause an error.
            this.set('layers', new LayerCollection());
        },_


Overriding initialize without calling the initialize from the superclass does break a lot :-)
I basically ended up with a useless situation, I couldn't send messages anymore.

But then again, I tried to find the correct syntax in this case, and the best I could find would be this:

MyCustomModel.\__super\__.initialize.call(this);

But that didn't work either, so I just moved the listenTo-call to the render method. 
It does what it needs for me, but I don't think it's the correct solution.

Evert
